File: #0239.sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {

        int size = nums.size();
        std::deque<pair<int, int>> deq;
        vector<int> result(size - k + 1);
        for (int i = 0; i < size; i++) {
            while ((deq.size() > 0) && (nums[i] > deq.back().first)) {
                deq.pop_back();
            }
            deq.push_back(make_pair(nums[i], i));
            
            if (deq.front().second == i - k) {
                deq.pop_front();
            }

            if (i >= k - 1) {
                result[i - (k - 1)] = deq.front().first;
            }
        }
        return result;
    }
};

File: #0239.sliding-window-maximum.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var maxSlidingWindow = function(nums, k) {
    let q = [];
    let result = [];
    for (let i = 0; i < nums.length; i++) {
        while (q.length && nums[i] >= q.at(-1).num) {
            q.pop();
        }

        q.push({idx: i, num: nums[i]});
        if (i >= k - 1) {
            while (q[0].idx <= i - k) {
                q.shift();
            }
            result.push(q[0].num);
        }
    }
    return result;
};
